bundles/config/org.eclipse.smarthome.config.discovery.test/src/test/groovy/org/eclipse/smarthome/config/discovery/DiscoveryServiceRegistryOSGiTest.java

```diff
@@ -0,0 +1,383 @@
+ * about the events fired by the according{@link DiscoveryService}.
+ * <p>
+ * This implementation creates two {@link DiscoveryService} mocks and registers
+ * them as service at the <i>OSGi</i> service registry. Since this test creates
+ * {@link DiscoveryResult}s which are added to the {@link Inbox},
+ * the {@link Inbox} is cleared again after this test returns.
+ *
+ * @author Michael Grammling - Initial Contribution
+ * @author Simon Kaufmann - added tests for ExtendedDiscoveryService, ported to Java
+ */
+public class DiscoveryServiceRegistryOSGiTest extends JavaOSGiTest {
+
+    private final String ANY_BINDING_ID_1 = "any2BindingId1";
+    private final String ANY_THING_TYPE_1 = "any2ThingType1";
+
+    private final String ANY_BINDING_ID_2 = "any2BindingId2";
+    private final String ANY_THING_TYPE_2 = "any2ThingType2";
+
+    private final String FAULTY_BINDING_ID = "faulty2BindingId";
+    private final String FAULTY_THING_TYPE = "faulty2ThingType";
+
+    private final String EXTENDED_BINDING_ID = "extended2BindingId";
+    private final String EXTENDED_THING_TYPE = "extended2ThingType";
+
+    private static class AnotherDiscoveryService extends DiscoveryServiceMock {
+        public AnotherDiscoveryService(ThingTypeUID thingType, int timeout) {
+            super(thingType, timeout);
+        }
+    }
+
+    private DiscoveryServiceMock discoveryServiceMockForBinding1;
+    private ExtendedDiscoveryServiceMock discoveryServiceMockForBinding2;
+    private DiscoveryServiceMock discoveryServiceFaultyMock;
+    private ExtendedDiscoveryServiceMock extendedDiscoveryServiceMock;
+    private DiscoveryServiceRegistry discoveryServiceRegistry;
+    private final List<ServiceRegistration<?>> serviceRegs = new ArrayList<>();
+    private ThingRegistry thingRegistry;
+    private Inbox inbox;
+
+    @Mock
+    private DiscoveryListener mockDiscoveryListener;
+
+    @Before
+    public void setUp() {
+        initMocks(this);
+        registerVolatileStorageService();
+
+        thingRegistry = getService(ThingRegistry.class);
+        assertNotNull(thingRegistry);
+
+        inbox = getService(Inbox.class);
+        assertNotNull(inbox);
+
+        discoveryServiceMockForBinding1 = new DiscoveryServiceMock(
+                new ThingTypeUID(ANY_BINDING_ID_1, ANY_THING_TYPE_1), 1);
+        discoveryServiceMockForBinding2 = new ExtendedDiscoveryServiceMock(
+                new ThingTypeUID(ANY_BINDING_ID_2, ANY_THING_TYPE_2), 3);
+
+        discoveryServiceFaultyMock = new DiscoveryServiceMock(new ThingTypeUID(FAULTY_BINDING_ID, FAULTY_THING_TYPE),
+                1, true);
+
+        extendedDiscoveryServiceMock = new ExtendedDiscoveryServiceMock(
+                new ThingTypeUID(EXTENDED_BINDING_ID, EXTENDED_THING_TYPE), 1, true);
+
+        serviceRegs.add(
+                bundleContext.registerService(DiscoveryService.class.getName(), discoveryServiceMockForBinding1, null));
```

Comment:
@afuechsel this was the nitty gritty detail: the JavaOSGiTest only remembers one service registration per interface hence didn't cleanup properly (I might as well fix _that_, but not tonight anymore...). So no matter if groovy or java runs first, the second one is always gonna fail because of the junk services.